### ISSUE TITLE
DeleteFromQuery repeated

### ISSUE THREAD

**Issue by NicolasGauvreau** [zzzprojects/EntityFramework-Extensions#154]
Greetings,

I have a logs table containing 101 entries and I do something similar to:
```
                    IQueryable<Log> logsToRemove = dbSet.Logs
                        .OrderBy(g => g.CreationDateTime)
                        .Take(51);

                    logsToRemove.DeleteFromQuery();
```
using the MSSQL  provider. I would expect the final table to contain only the latest 50 entries.
However, after the DeleteFromQuery, my table is empty.

When I look at the SQL Server Profiler, I can see this:
```
MERGE INTO [SCHEMANAME].[LOGS] AS DestinationTable
USING
(
SELECT TOP (51) 
    [Extent1].[ID] AS [ID]
    FROM [SCHEMANAME].[LOGS] AS [Extent1]
    ORDER BY [Extent1].[CREATIONDATETIME] ASC
) AS StagingTable
ON StagingTable.ID = DestinationTable.ID
WHEN MATCHED THEN
    DELETE
;
go
MERGE INTO [SCHEMANAME].[LOGS] AS DestinationTable
USING
(
SELECT TOP (51) 
    [Extent1].[ID] AS [ID]
    FROM [SCHEMANAME].[LOGS] AS [Extent1]
    ORDER BY [Extent1].[CREATIONDATETIME] ASC
) AS StagingTable
ON StagingTable.ID = DestinationTable.ID
WHEN MATCHED THEN
    DELETE
;
go
MERGE INTO [SCHEMANAME].[LOGS] AS DestinationTable
USING
(
SELECT TOP (51) 
    [Extent1].[ID] AS [ID]
    FROM [SCHEMANAME].[LOGS] AS [Extent1]
    ORDER BY [Extent1].[CREATIONDATETIME] ASC
) AS StagingTable
ON StagingTable.ID = DestinationTable.ID
WHEN MATCHED THEN
    DELETE
;
go
```
The query is OK, but is sent 3 times (which explains the empty table).
Can you explain to me what can cause the query to be sent 3 times?

Thanks,

Nicolas

**Comment by JonathanMagnan** [comment]
Hello @NicolasGauvreau ,

I believe the `Take` statement is not supported.

The way we created this feature is to execute a `DELETE` until there are no more rows matching the criteria to delete.

Obviously, this strategy doesn't work with your case when a `Take` statement is used.

We already have planned to optimize this kind of statement in our roadmap (Will be started next month) to make only one database round-trip and use `DELETE` statement directly whenever we can. We will make sure to support that scenario as well.

Let me know if that answer correctly to your issue until it gets fixed.

Best Regards,

Jonathan

**Comment by NicolasGauvreau** [comment]
Thanks for the feedback.

For your behavior: “execute a DELETE until there are no more rows matching the criteria to delete”, it would have been nice if it could have simply been an optional switch which can be turned off, e.g.:

            query.DeleteFromQuery(options =>
            {
                options.ReattemptUntilNoneAffected = false; // Delete query will be executed only once
            });

I’ll refrain from using “DeleteFromQuery” with a “Take” statement for the time being.

Thanks again,

Nicolas

**Comment by JonathanMagnan** [comment]
Thank you @NicolasGauvreau for your suggestion.

You are right, we should look to add something like this.

We will look at this issue and choose if we permanently fix (best scenario) or add a similar option temporary.
